packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -173,43 +206,62 @@ export class ConnectionStateHandler {
         // ops sent by this client, so we should keep the old client id until we see our own client's
-        // join message. after we see the join message for out new connection with our new client id,
+        // join message. after we see the join message for our new connection with our new client id,
         // we know there can no longer be outstanding ops that we sent with the previous client id.
         this._pendingClientId = details.clientId;
 
-        // Report telemetry after we set client id, but before transitioning to Connected state below!
+        // This pending clientId could be in the quorum already (i.e. join op already processed).
+        // We are fetching ops from storage in parallel to connecting to Relay Service,
+        // and given async processes, it's possible that we have already processed our own join message before
+        // connection was fully established.
+        // If quorumClients itself is undefined, we expect it will process the join op after it's initialized.
+        const waitingForJoinOp = quorumClients?.getMember(this._pendingClientId) === undefined;
+
+        // IMPORTANT: Report telemetry after we set _pendingClientId, but before transitioning to Connected state
         this.handler.logConnectionStateChangeTelemetry(ConnectionState.Connecting, oldState);
 
-        const quorumClients = this.handler.quorumClients();
-        // Check if we already processed our own join op through delta storage!
-        // we are fetching ops from storage in parallel to connecting to ordering service
-        // Given async processes, it's possible that we have already processed our own join message before
-        // connection was fully established.
-        // Note that we might be still initializing quorum - connection is established proactively on load!
-        if (quorumClients?.getMember(details.clientId) !== undefined
-            || connectionMode === "read"
-        ) {
-            assert(!this.prevClientLeftTimer.hasTimer, 0x2a6 /* "there should be no timer for 'read' connections" */);
-            this.setConnectionState(ConnectionState.Connected);
-        } else if (connectionMode === "write") {
+        if (writeConnection && waitingForJoinOp) {
+            // Previous client left, and we are waiting for our own join op. When it is processed we'll join the quorum
+            // and attempt to transition to Connected state via receivedAddMemberEvent.
             this.startJoinOpTimer();
+        } else if (this.waitingForLeaveOp) {
+            // Previous client left, and we are waiting for its leave op.
+            // Nothing to do now - when the previous client is removed from the quorum
+            // we will attempt to transition to Connected state via receivedRemoveMemberEvent
+            assert(writeConnection, 0x2a6 /* "there should be no timer for 'read' connections" */);
```

Comment:
Can this checked be moved out of these IF-else block and be like that?

assert(!this.waitingForLeaveOp || writeConnection, 0x2a6 /* "there should be no timer for 'read' connections" */);